docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+direction towards the sinks is **downward**, and the direction towards the
+sources is **upward**.
+
+#### Parameters
+
+In our example, we want all of our circuitry to agree upon a Scala `Int` to use
+as a wire width. For clarity, we will define distinct case classes for our
+parameters, although they all contain the same type.
+
+```scala mdoc
+case class UpwardParam(width: Int)
+case class DownwardParam(width: Int)
+case class EdgeParam(width: Int)
+```
+
+#### Node Implementation
+
+In our "node implementation", or `NodeImp`, we describe how parameters flow
+throughout our graph and how parameters are negotiated between nodes.
+
+We will use `SimpleNodeImp`, which performs the same parameter negotiation and
+passes the same bundles along an edge, regardless of whether the edge points
+into our out of the node.
+
+Our **edge parameter** (`E`) describes the data type that needs to be passed along
+the edges in our graph; in our case we only need to keep track of one `Int` that
+describes our width, which will live inside our `EdgeParam` class.
+
+We also specify a **bundle parameter** (`B`), which is the type of data that will
+resolve into parameterized hardware ports between our modules. For us, this is
+a Chisel `UInt` with the width described by our edge parameter (see the `bundle`
+function).
+
+```scala mdoc
+// PARAMETER TYPES:                       D              U            E          B
+object AdderNodeImp extends SimpleNodeImp[DownwardParam, UpwardParam, EdgeParam, UInt] {
+  def edge(pd: DownwardParam, pu: UpwardParam, p: Parameters, sourceInfo: SourceInfo) = {
+    if (pd.width < pu.width) EdgeParam(pd.width) else EdgeParam(pu.width)
+  }
+  def bundle(e: EdgeParam) = UInt(e.width.W)
+  def render(e: EdgeParam) = RenderedEdge("blue", s"width = ${e.width}")
+}
+```
+
+Notice that the `edge` function does the actual negotiation between nodes. In our
+case, it compares the parameters traveling both upwards and downwards, and chooses
+the one with the smaller width to pass along.
+
+The `render` function is required by `NodeImp`s and holds metadata for graph
+generation.
+
+#### Nodes
+
+Nodes can receive or generate parameters. A module parameterized through
+Diplomacy must have one or more nodes in order to receive or send parameters.
+Here, we will create our nodes for the modules we would like to parameterize.
+
+Let's start off with creating a node for our driver. We will make it a
+`SourceNode` with the node implementation shown in the previous section, since
+`SourceNode`s have only outward pointing edges.
+
+For our `AdderDriverNode`, `widths` of type `Seq[DownwardParam]` represents
+the outputs of this node. We are using a `Seq` here because each node will send
+two identical outputs: one to the adder, and the other to the monitor. These
+properties are `require`d by the node.
+
+```scala mdoc
+/** node for [[AdderDriver]] (source) */
+class AdderDriverNode(widths: Seq[DownwardParam])(implicit valName: ValName)
```

Comment:
In real code I would make you take a single `width: Int` argument, and then pass `Seq.fill(2){ DownwardParam(width) }` to `SourceNode`. Then there is is no way that the user can ever get it wrong, and so you don't even need the `require` checks! But do you think that would be helping or hurting the example?

Reply:
I don't think that would hurt the example. But for type-checking reasons in the NodeImp, wouldn't I need to do pass in `width: DownwardParam` instead of `Int`? Like so:

```scala
class AdderDriverNode(width: DownwardParam)(implicit valName: ValName)
  extends SourceNode(AdderNodeImp)(Seq(width))
```

and then I could do `Seq.fill(2){ width }`. Or is there some other way to go about this?

Reply:
Ohhh ignore my previous comment, I see what you are doing now. I think that is a neat way of doing it, but a but hard to parse for diplomacy newbies. I think it could be useful to include both versions, and explain how they are doing the same thing.

Reply:
Actually upon more consideration, I think your way of doing it is better to show.

Reply:
Yeah I think we are building up an example that shows lots of different ways of doing things in different cases, which, while it makes the example code less uniform, has the benefit of suggesting those different options as available to readers